flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricStore.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.runtime.webmonitor.metrics;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Nested data-structure to store metrics.
+ *
+ * This structure is not thread-safe.
+ */
+public class MetricStore {
+	private static final Logger LOG = LoggerFactory.getLogger(MetricStore.class);
+
+	JobManagerMetricStore jobManager = new JobManagerMetricStore();
+	Map<String, TaskManagerMetricStore> taskManagers = new HashMap<>();
+	Map<String, JobMetricStore> jobs = new HashMap<>();
+
+	/**
+	 * Adds a metric to this MetricStore.
+	 *
+	 * @param name  the metric identifier
+	 * @param value the metric value
+	 */
+	public void add(String name, Object value) {
```

Comment:
To be honest, I'm not a very big fan of encoding type and hierarchical information in a string which has to be reparsed in order to reconstruct the afore-mentioned information. The problem with this approach is that everything is very implicit and you don't have a tight coupling (in terms of format) between the sender and receiver. If something changes at the sender side, you won't notice at all that you have to change here something as well. Even at runtime the only thing you see is that you don't see the metrics. This makes it very hard to debug. I would be in favour of creating for the different types different messages, e.g. `JobManagerMetric`, `TaskManagerMetric`, `TaskMetric`, etc. These messages contain the respective information. Furthermore, for the histogram type case, you will have to do the whole parsing over and over again instead of doing it once.